### PR TITLE
Issue #779: added to_number() and to_str()

### PR DESCRIPTION
**Pull request by rafie**
Also added error reporting to RSValue_Cmp() via QueryError.
When error reporting argis not provided, RSValue_Cmp will fallback to string comparison.

### CODE REVIEW

File: src/aggregate/functions/string.c [hunk 1]
```diff
@@ -9,6 +9,8 @@
 
 #define STRING_BLOCK_SIZE 512
 
+///////////////////////////////////////////////////////////////////////////////////////////////
```
[the remark this answers is not in the document]
Comment (rafie) [reply src/aggregate/functions/string.c]: I find that function-separation comments improve readability, but it is possible that it caters to my own uncomfort with large chunks of text being thrown at me. :)
It it looks too noisy for you, I'll remote them.

File: src/aggregate/functions/string.c [hunk 2]
```diff
@@ -121,6 +129,36 @@ static int stringfunc_substr(ExprEval *ctx, RSValue *result, RSValue **argv, siz
...
+//---------------------------------------------------------------------------------------------
+
+int func_to_number(ExprEval *ctx, RSValue *result, RSValue **argv, size_t argc,
+                   QueryError *err) {
+  VALIDATE_ARGS("to_number", 1, 1, err);
+
+  double n;
+  if (!RSValue_ToNumber(argv[0], &n)) {
+    size_t sz = 0;
+    char *p = (char *)RSValue_StringPtrLen(argv[0], &sz);
```
[the remark this answers is not in the document]
Comment (rafie) [reply src/aggregate/functions/string.c]: True. Fixed.

File: src/cpptests/t_index.cpp [hunk 3]
```diff
@@ -1058,21 +1058,24 @@ TEST_F(IndexTest, testSortable) {
...
+
+  int rc = RSSortingVector_Cmp(v, v2, &sk, &qerr);
+  ASSERT_TRUE(rc > 0 && qerr.code == QUERY_OK);
   sk.ascending = 1;
-  rc = RSSortingVector_Cmp(v, v2, &sk);
-  ASSERT_TRUE(rc < 0);
-  rc = RSSortingVector_Cmp(v, v, &sk);
-  ASSERT_EQ(0, rc);
+  rc = RSSortingVector_Cmp(v, v2, &sk, &qerr);
+  ASSERT_TRUE(rc < 0 && qerr.code == QUERY_OK);
```
[the remark this answers is not in the document]
Comment (rafie) [reply src/cpptests/t_index.cpp]: Done.

**Comment by rafie** [comment]
I'd like us to also consider the new function names: to_number(), to_str(), as once they are introduced there is no going back.
Alternatives that come to mind: to_num(), number(), string(), toNumber(), toString().

File: src/aggregate/expr/expression.c [hunk 4]
```diff
@@ -123,6 +124,9 @@ static int getPredicateBoolean(const RSValue *l, const RSValue *r, RSCondition o
 
     default:
       assert("Unknown predicate received" && 0);
+      if (qerr) {
+        QueryError_SetErrorFmt(qerr, QUERY_EEXPR, "Unknown predicate received");
+      }
```
Comment (MeirShpilraien) [inline src/aggregate/expr/expression.c]: @rafie I might have missed something but we have an assert(0) before, this is a dead code.

File: src/aggregate/expr/expression.c [hunk 5]
```diff
@@ -152,11 +155,14 @@ static int evalPredicate(ExprEval *eval, const RSPredicate *pred, RSValue *resul
   RSValue *l_ptr = RSValue_Dereference(&l);
   RSValue *r_ptr = RSValue_Dereference(&r);
 
-  int res = getPredicateBoolean(&l, &r, pred->cond);
-
-  result->numval = res;
-  result->t = RSValue_Number;
-  rc = EXPR_EVAL_OK;
+  int res = getPredicateBoolean(eval, &l, &r, pred->cond);
+  if (!eval->err || eval->err->code == QUERY_OK) {
```
Comment (MeirShpilraien) [inline src/aggregate/expr/expression.c]: Wouldn't this might change the behavior?

File: src/value.c [hunk 6]
```diff
@@ -473,7 +476,25 @@ static inline int cmp_strings(const char *s1, const char *s2, size_t l1, size_t
   }
 }
 
-int RSValue_Cmp(const RSValue *v1, const RSValue *v2) {
+static inline int cmp_numbers(const RSValue *v1, const RSValue *v2) {
+  return v1->numval > v2->numval ? 1 : (v1->numval < v2->numval ? -1 : 0);
+}
+
+static inline int convert_to_numeber(const RSValue *v, RSValue *vn, QueryError *qerr) {
```
Comment (MeirShpilraien) [inline src/value.c]: type : number

File: src/numeric_filter.c [hunk 7]
```diff
@@ -72,6 +72,9 @@ NumericFilter *NumericFilter_Parse(ArgsCursor *ac, QueryError *status) {
 }
 
 void NumericFilter_Free(NumericFilter *nf) {
+  if (!nf) {
```
Comment (MeirShpilraien) [inline src/numeric_filter.c]: How can we get a NULL filter ?

Comment (rafie) [reply src/numeric_filter.c]: This fixes an actual crash that I had. I don't remember how that came to be.

File: src/aggregate/expr/expression.c (hunk 5 above)
Comment (rafie) [reply src/aggregate/expr/expression.c]: It probably will, but I don't see any other way to report an error coming out of getPredicateBoolean(). If we don't return EXPR_EVAL_ERR, the error will disappear, and if we don't set RSValue_Undef, we've returned a number that isn't there.

File: src/aggregate/expr/expression.c (hunk 4 above)
Comment (rafie) [reply src/aggregate/expr/expression.c]: Right.

File: src/value.c (hunk 6 above)
Comment (rafie) [reply src/value.c]: Please explain.

Comment (MeirShpilraien) [reply src/value.c]: numeber -> number :)

Comment (rafie) [reply src/value.c]: My eyes are getting old. :-)

File: src/aggregate/functions/string.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (mnunberg) [reply src/aggregate/functions/string.c]: @rafie I have my own script for this, whih is arguably cleaner.. works rather nicely

https://github.com/mnunberg/mkcomment/blob/master/mkcomment.pl

[the remark this answers is not in the document]
Comment (rafie) [reply src/aggregate/functions/string.c]: Comments removed.

File: src/value.c [hunk 8]
```diff
@@ -517,8 +562,66 @@ int RSValue_Cmp(const RSValue *v1, const RSValue *v2) {
...
 
-int RSValue_Equal(const RSValue *v1, const RSValue *v2) {
-  return RSValue_Cmp(v1, v2) == 0;
+int RSValue_Equal(const RSValue *v1, const RSValue *v2, QueryError *qerr) {
+  assert(v1);
+  assert(v2);
+  v1 = RSValue_Dereference(v1);
+  v2 = RSValue_Dereference(v2);
+
+  if (v1->t == v2->t) {
```
Comment (mnunberg) [inline src/value.c]: @rafie Why don't you just call RSValue_Cmp as before? seems like needless duplication

**Pull request merged**